client/ayon_core/addon/utils.py

```diff
@@ -150,35 +148,52 @@ def ensure_addons_are_process_context_ready(
         detail = output_str
 
     _handle_error(process_context, message, detail)
-    if not exit_on_failure:
-        return exception
-    sys.exit(1)
+    if exit_on_failure:
+        sys.exit(1)
 
 
 def ensure_addons_are_process_ready(
+    addon_name: Optional[str] = None,
+    addon_version: Optional[str] = None,
+    project_name: Optional[str] = None,
+    headless: Optional[bool] = None,
     addons_manager: Optional[AddonsManager] = None,
     exit_on_failure: bool = True,
     **kwargs,
-) -> Optional[Exception]:
+) -> ProcessContext:
     """Ensure all enabled addons are ready to be used in the given context.
 
     Call this method only in AYON launcher process and as first thing
         to avoid possible clashes with preparation. For example 'QApplication'
         should not be created.
 
     Args:
+        addon_name (Optional[str]): Addon name which triggered process.
+        addon_version (Optional[str]): Addon version which triggered process.
+        project_name (Optional[str]): Project name. Can be filled in case
+            process is triggered for specific project. Some addons can have
+            different behavior based on project. Value is NOT autofilled.
+        headless (Optional[bool]): Is process running in headless mode. Value
+            is filled with value based on state set in AYON launcher.
         addons_manager (Optional[AddonsManager]): The addons
             manager to use. If not provided, a new one will be created.
         exit_on_failure (bool, optional): If True, the process will exit
             if an error occurs. Defaults to True.
         kwargs: The keyword arguments to pass to the ProcessContext.
 
     Returns:
-        Optional[Exception]: The exception that occurred during the
+        ProcessContext: The exception that occurred during the
             preparation, if any.
 
     """
-    context: ProcessContext = ProcessContext(**kwargs)
-    return ensure_addons_are_process_context_ready(
+    context: ProcessContext = ProcessContext(
+        addon_name,
+        addon_version,
+        project_name,
+        headless,
+        **kwargs
+    )
+    ensure_addons_are_process_context_ready(
         context, addons_manager, exit_on_failure
     )
+    return context
```

Comment:
Does whoever receive this object have any meaningful reason to set the exception and or set the prepared state - I assume not? I have a feeling if maybe we should just be returning a (static?) data object instead to avoid confusion? We're really just returning the prepared state and the exception - that's it?

Reply:
For my usecase the context has more value as it fills `headless` based on logic I should not know about at that specific case. I did want to be able to get exception to be able to handle the error in custom way.

For me `ensure_addons_are_process_ready` is just helper wrapper for exactly the content of the function when I don't want to encapsulate the logic to e.g. UI splash screen, but only run preparation and then get context information that was filled automatically.

Reply:
Hmm - I see. I suppose the change is made to fit your use case and there are reasons to do it as such.

Reply:
Unresolving. Those are relevant comments and questions. This is not yet in any release and I would rather not touch it in future, so criticism is welcome.

Full context:
The main reason this functionality is necessary is to give addons option to prepare them -> what would happen in tray if you would use it, like fill ftrack credentials, fill environment variables that are filled only in `tray_start` etc. They can stop the process if their need is not matched. Main usage is for webactions, they can be started without ever running tray.

This cannot happen automatically, addon that might need it has to request it specifically.

Function `ensure_addons_are_process_ready` was added to simplify
```python
from ayon_core.addon import ProcessContext, ensure_addons_are_process_context_ready

def some_func():
    context = ProcessContext(
        "myaddon",
        "1.1.1",
        "myProject"
    )
    ensure_addons_are_process_context_ready(context)
    if context.headless:
        _headless_stuff()
    else:
        _ui_stuff()
```


```python
from ayon_core.addon import ensure_addons_are_process_ready

def some_func():
    context = ensure_addons_are_process_ready(
        "myaddon",
        "1.1.1",
        "myProject"
    )
    if context.headless:
        _headless_stuff()
    else:
        _ui_stuff()
```

Reply:
I'm still not entirely sure about the exact use case. I think _if_ we're simplifying it makes more sense to simplify the need to pass the addon version (isn't that e.g. based on the current active bundle for your context?) Why do we need to explicitly pass it the version?

The fact that you need access to any exceptions or whether it's headless or not - I suppose makes sense. I'm also not entirely confident about the need for the `prepared` state.

I'd expect a `ensure_addons_are_process_ready` to just return something along the lines of a "boolean" to tell me whether it succeeded or not, potentially with the exception if it's not (which could also be an actual raised exception? Since we're requesting it to _ensure_ it?)

Again, take my comment with a grain of salt since I feel I'm quite on a different track than you and still not 100% sure.

Just here proposing an API that may seem more reasonable:
```python
context = ProcessContext(project_name="myProject")
result = context.initialize_for_addon("myaddon")
if not result:
    print(result.exception)

if context.is_headless:
    _headless_stuff()
else:
    _ui_stuff()
```

But maybe it's the docstring of `ProcessContext` that needs improving - because it doesn't seem to be 'generic context' but a process context specific to a particular addon? So it's more of a `AddonProcessContext`?

Reply:
First idea was to expect `addon_name`, `addon_version` and `project_name` in `ensure_addons_are_process_ready`. Then I realized that we will in future need to add more data and changing signature of a function is PITE, thus `ProcessContext` was created, meant as "future-proof data holder".